core/templates/dev/head/domain/collection/CollectionValidationService.js

```diff
@@ -0,0 +1,171 @@
+      var prerequisiteSkillList = collectionNode.getPrerequisiteSkillList();
+      var acquiredSkillList = collectionNode.getAcquiredSkillList();
+      var overlappingSkillList = SkillListObjectFactory.create([]);
+      prerequisiteSkillList.getSkills().forEach(function(skill) {
+        if (acquiredSkillList.containsSkill(skill)) {
+          overlappingSkillList.addSkill(skill);
+        }
+      });
+      acquiredSkillList.getSkills().forEach(function(skill) {
+        if (prerequisiteSkillList.containsSkill(skill)) {
+          overlappingSkillList.addSkill(skill);
+        }
+      });
+      return overlappingSkillList.getSkills();
+    };
+
+    var _getNextExplorationIds = function(collection, completedExpIds) {
+      var acquiredSkillList = completedExpIds.reduce(
+        function(skillList, explorationId) {
+          var collectionNode = collection.getCollectionNodeByExplorationId(
+            explorationId);
+          skillList.addSkillsFromSkillList(
+            collectionNode.getAcquiredSkillList());
+          return skillList;
+        }, SkillListObjectFactory.create([]));
+
+      // Pick all collection nodes whose prerequisite skills are satisified by
+      // the currently acquired skills and which has not yet been completed.
+      return collection.getExplorationIds().filter(function(explorationId) {
+        var collectionNode = collection.getCollectionNodeByExplorationId(
+          explorationId);
+        return completedExpIds.indexOf(explorationId) == -1 &&
+          acquiredSkillList.isSupersetOfSkillList(
+            collectionNode.getPrerequisiteSkillList());
+      });
+    };
+
+    var _getUnreachableExplorationIds = function(collection) {
+      var completedExpIds = _getStartingExplorationIds(collection);
+      var nextExpIds = _getNextExplorationIds(collection, completedExpIds);
+      while (nextExpIds.length > 0) {
+        completedExpIds = completedExpIds.concat(nextExpIds);
+        nextExpIds = _getNextExplorationIds(collection, completedExpIds);
+      }
+
+      return collection.getExplorationIds().filter(function(explorationId) {
+        return completedExpIds.indexOf(explorationId) == -1;
+      });
+    };
+
+    var _getNonexistentExplorationIds = function(collection) {
+      return collection.getCollectionNodes().filter(function(collectionNode) {
+        return !collectionNode.doesExplorationExist();
+      }).map(function(collectionNode) {
+        return collectionNode.getExplorationId();
+      });
+    };
+
+    var _getPrivateExplorationIds = function(collection) {
+      return collection.getCollectionNodes().filter(function(collectionNode) {
+        return collectionNode.isExplorationPrivate();
+      }).map(function(collectionNode) {
+        return collectionNode.getExplorationId();
+      });
+    };
+
+    var _validateCollection = function(collection, isPublic) {
+      // NOTE(bhenning): These validation issues should be similar to those
```

Comment:
I don't understand -- is this a TODO? If not, remove NOTE(bhenning) and replace with NOTE TO DEVELOPERS.

Also, maybe make it more explicit: "Please ensure that this validation logic is the same as that in core.domain.collection_domain.Collection.validate()". You might want to make a corresponding note in core.domain.collection_domain.Collection.validate(), too, as a reminder to developers to keep both of these in sync.